cpp/ql/src/semmle/code/cpp/rangeanalysis/SimpleRangeAnalysis.qll

```diff
@@ -156,8 +156,65 @@ float safeFloor(float v) {
   result = v
 }
 
+/** A `MulExpr` where exactly one operand is constant. */
+private class MulByConstantExpr extends MulExpr {
+  float constant;
+  Expr operand;
+
+  MulByConstantExpr() {
+    exists(Expr constantExpr |
+      this.hasOperands(constantExpr, operand) and
+      constant = constantExpr.getFullyConverted().getValue().toFloat() and
+      not exists(operand.getFullyConverted().getValue().toFloat())
+    )
+  }
+
+  /** Gets the value of the constant operand. */
+  float getConstant() { result = constant }
+
+  /** Gets the non-constant operand. */
+  Expr getOperand() { result = operand }
+}
+
 private class UnsignedMulExpr extends MulExpr {
-  UnsignedMulExpr() { this.getType().(IntegralType).isUnsigned() }
+  UnsignedMulExpr() {
+    this.getType().(IntegralType).isUnsigned() and
+    // Avoid overlap. It should be slightly cheaper to analyze
+    // `MulByConstantExpr`.
+    not this instanceof MulByConstantExpr
+  }
+}
+
+/**
+ * Holds if `expr` is effectively a multiplication of `operand` with the
+ * positive constant `positive`.
+ */
+private predicate multipliesByPositive(Expr expr, Expr operand, float positive) {
+  operand = expr.(MulByConstantExpr).getOperand() and
+  positive = expr.(MulByConstantExpr).getConstant() and
+  positive >= 0.0 // includes positive zero
+  or
+  operand = expr.(UnaryPlusExpr).getOperand() and
+  positive = 1.0
+  or
+  operand = expr.(CommaExpr).getRightOperand() and
+  positive = 1.0
+  or
+  operand = expr.(StmtExpr).getResultExpr() and
+  positive = 1.0
```

Comment:
I get that there's nothing technically wrong with including these cases in this predicate, but it seems wrong to have these non-multiplication expressions in a predicate named `multipliesByPositive`. What's the motivation for including these cases?

Reply:
I originally used the name `effectivelyMultipliesByPositive` but found it a bit long. Would that be better?

The motivation is to take as many cases as possible out of the recursive predicates and move them into this non-recursive predicate. I'm adding so many extra recursive cases, so I thought it would be best for readability and performance if I remove some where I can. I haven't observed any performance improvement.

Reply:
That makes sense. I think I prefer the longer name as it doesn't make it sound like the `expr` must be a `MulExpr`.